### PR TITLE
Allow ref structs to implement interfaces, respect "allows ref struct" during constraints check.

### PR DESCRIPTION
**Pull request by AlekseyTs**

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -7917,4 +7914,13 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_ClassIsCombinedWithRefStruct" xml:space="preserve">
     <value>Cannot allow ref structs for a type parameter known from other constraints to be a class</value>
   </data>
+  <data name="ERR_NotRefStructConstraintNotSatisfied" xml:space="preserve">
+    <value>The type '{2}' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter '{1}' in the generic type or method '{0}'</value>
```
Comment (cston) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:7918]: >The type '{2}' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter '{1}' in the generic type or method '{0}' [](http://example.com/codeflow?start=11&length=153)

Perhaps: "The type ... **must not be** a ref struct ... in order to use it as **type** parameter ..."

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs [hunk 2]
```diff
@@ -5316,5 +5316,846 @@ void verify(ModuleSymbol m)
...
+        return x;
+    }
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            // PROTOTYPE(RefStructInterfaces): The boxing should be disallowed.
+            comp.VerifyDiagnostics(
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:6136]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing explicit cast as well: `return (object)x;`.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs [hunk 3]
```diff
@@ -5316,5 +5316,846 @@ void verify(ModuleSymbol m)
...
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:6159]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing explicit cast as well: `return (I1)x;`.

Also:
```csharp
static U Test1<T, U>(T x)
    where T : allows ref struct
    where U : T
{
    return x;
}
static U Test2<T, U>(T x)
    where T : allows ref struct
    where U : T
{
    return (U)x;
}
```

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs [hunk 4]
```diff
@@ -5316,5 +5316,846 @@ void verify(ModuleSymbol m)
...
+}
+";
+            var comp = CreateCompilation(src, targetFramework: s_targetFrameworkSupportingByRefLikeGenerics);
+
+            comp.VerifyDiagnostics(
+                // (12,26): error CS9504: The type 'S1' may not be a ref struct or a type parameter allowing ref structs in order to use it as parameter 'S' in the generic type or method 'C<T, S>'
+                //         _ = typeof(C<S1, S1>);
+                Diagnostic(ErrorCode.ERR_NotRefStructConstraintNotSatisfied, "S1").WithArguments("C<T, S>", "S", "S1").WithLocation(12, 26)
+                );
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:6117]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing:
```csharp
_ = typeof(C<T, T>);
_ = typeof(C<S, S>);
```

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs [hunk 5]
```diff
@@ -5316,5 +5316,846 @@ void verify(ModuleSymbol m)
...
+  // Code size       15 (0xf)
+  .maxstack  2
+  IL_0000:  ldarga.s   V_0
+  IL_0002:  ldarg.1
+  IL_0003:  constrained. ""T""
+  IL_0009:  callvirt   ""void I1<string>.M1(string)""
+  IL_000e:  ret
+}
+");
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:5516]: >} [](http://example.com/codeflow?start=8&length=1)

Consider testing variance error cases as well:
```csharp
interface I<T>
{
    void M(T t);
}
interface IOut<out T>
{
    T MOut();
}
ref struct S : I<object>, IOut<object>
{
    public void M(object o) { }
    public object MOut() => null;
}
class Program
{
    static void Main()
    {
        Test1(new S());
        Test2(new S());
    }
    static void Test1<T>(T x) where T : I<string>, allows ref struct
    {
    }
    static void Test2<T>(T x) where T : IOut<string>, allows ref struct
    {
    }
}```

File: src/Compilers/CSharp/Portable/CSharpResources.resx (hunk 1 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/CSharpResources.resx:7918]: For "may not be a", I think I used the following message as a template:
```
  <data name="ERR_LambdaInIsAs" xml:space="preserve">
    <value>The first operand of an 'is' or 'as' operator may not be a lambda expression, anonymous method, or method group.</value>
  </data>
```

For "in order to use it as parameter", there is a "prior art" too:
```
  <data name="ERR_RefConstraintNotSatisfied" xml:space="preserve">
    <value>The type '{2}' must be a reference type in order to use it as parameter '{1}' in the generic type or method '{0}'</value>
  </data>
```

I can make the suggested change if you still think it is needed. Please let me know.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs (hunk 2 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:6136]: > Are we testing boxing for this?

Why would this make a difference? Conversion from ref struct to an interface is covered in ```ImplementAnInterface_02_IllegalBoxing```. I do not find anything interesting in the suggested scenario. If I am missing something, please let me know.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs (hunk 3 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:6159]: > Consider testing explicit cast as well: return (I1)x;.

Testing boxing is not a goal of this PR. It is known that proper checks are not in place and the added tests a sufficient to make the knowledge obvious

**Review by cston — APPROVED** [review]

**Comment by AlekseyTs** [comment]
@jjonescz, @dotnet/roslyn-compiler For the second review.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs [hunk 6]
```diff
@@ -5316,5 +5316,937 @@ void verify(ModuleSymbol m)
...
+
+            CompileAndVerify(comp, sourceSymbolValidator: verify, symbolValidator: verify).VerifyDiagnostics();
+
+            void verify(ModuleSymbol m)
+            {
+                var s1 = m.GlobalNamespace.GetMember<NamedTypeSymbol>("S1");
+                Assert.Equal("I1", s1.InterfacesNoUseSiteDiagnostics().Single().ToTestDisplayString());
+            }
+
+            CreateCompilation(src, targetFramework: TargetFramework.Net80, parseOptions: TestOptions.RegularNext).VerifyDiagnostics();
```
Comment (jjonescz) [inline src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:5340]: Should we also test `RegularPreview` to ensure this keeps working in C#14+?

**Review by jjonescz — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs [hunk 7]
```diff
@@ -1307,13 +1331,20 @@ static ArrayBuilder<TypeParameterDiagnosticInfo> ensureUseSiteDiagnosticsBuilder
...
-                return true;
+                // "... A boxing conversion (6.1.7), provided that type A is a non-nullable value type. ..."
+                // NOTE: we extend this to allow, for example, a conversion from Nullable<T> to object.
+                if (conversions.HasBoxingConversion(typeArgument.Type.IsNullableType() ? ((NamedTypeSymbol)typeArgument.Type).ConstructedFrom : typeArgument.Type,
+                                                    constraintType.Type, ref useSiteInfo))
+                {
+                    return true;
+                }
+
+                if (typeArgument.Type.IsRefLikeType && conversions.ImplementsVarianceCompatibleInterface(typeArgument.Type, constraintType.Type, ref useSiteInfo))
```
[the remark this answers is not in the document]
Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:1344]: > Why is this check needed specifically for ref struct type arguments - i.e., why they need different handling then normal structs for example?

There are no boxing conversion for ref structs, therefore, the ```conversions.HasBoxingConversion``` check above fails.

File: src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs (hunk 6 above)
Comment (AlekseyTs) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:5340]: > Should we also test RegularPreview to ensure this keeps working in C# 14+?

It is tested, that is the default option for our tests. Therefore, the ```CompileAndVerify``` covers that

Comment (jjonescz) [reply src/Compilers/CSharp/Test/Emit2/Semantics/RefStructInterfacesTests.cs:5340]: I've overlooked that, thanks.

**Pull request merged**